서버 배포 및 cicd 구축

Jeongh00 opened the pull request:
AWS 에 EC2 , RDS 인스턴스 모두 생성 완료했고, Github Action 으로 자동 배포 까지 완료했습니다.
현재는, feature/infra-basic 을 기준으로 push 될떄 로 되어있습니다.

작업하면서 겪었던 문제들을 좀 공유하고자 합니다.
먼저, 멀티모듈에서의 main class 를 찾지 못하여 api module 로 설정해 주었습니다.
다음으로, BIT(1) type 으로 적어주면 syntax 에러가 났는데 저도 이부분은 정확히 파악하지 못했는데 혹시 알고 계신가요? 결론적으로는 BIT 타입으로 바꿨더니 잘 실행되었습니다.
세번째로, bootJar 가 false 로 되어있던 gradle 부분을 mainClass 설정 후 enabled true로 바꿔주었습니다.
마지막으로는 멀티 모듈 구조에 맞게 settings.gradle 파일도 설정해 주었습니다.

후에 다시 기준을 main branch 로 바꿀 예정이고, merge 하면서 테스트 하면 좋을것같습니다^^

Jeongh00 committed:
chore: mainclass 지정 & settings 파일 설정

Jeongh00 committed:
feat: dockerfile 생성 & compose.yml 파일 작성

Jeongh00 committed:
fix: bit type update & compose.yml 파일 수정

Jeongh00 committed:
Create cd.yml

Jeongh00 committed:
test: test for cd

Jeongh00 committed:
test: test

Jeongh00 committed:
fix: jdk version 설정

Jeongh00 committed:
fix: 권한 위해 sudo 명령어 추가

Jeongh00 committed:
fix: 불필요 의존성 삭제

ldk980130 commented on .github/workflows/cd.yml:
```diff
@@ -0,0 +1,43 @@
+  push:
+    branches: [ feature/infra-basic ] # 임시로 브랜치 설정
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up JDK 17
+        uses: actions/setup-java@v3
+        with:
+          java-version: 17
+          distribution: 'adopt'
+
+      - name: 저장소 Checkout
+        uses: actions/checkout@v2
+
+      - name: 스프링부트 애플리케이션 빌드
+        run: ./gradlew bootJar
```
빌드하기 전에 서브모듈을 업데이트하는 코드를 넣어줘야 해요! 이렇게 하지 않으면 서브모듈을 인식 못하거나 옛날 커밋을 바라보는 현상이 생기더라구요
``` yml
jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - name: 저장소 Checkout
        uses: actions/checkout@v3
        with:
          submodules: true
          token: ${{ secrets.ACTION_TOKEN }}

      - name: update submodules
        run: git submodule update --remote
```

Jeongh00 replied:
아 그렇네요.. submodule을 생각치 못했네요 반여하겠습니다!

Jeongh00 replied:
token 생성 후 설정하였습니다^^.

ldk980130 commented on Dockerfile:
```diff
@@ -0,0 +1,5 @@
+ARG JAR_FILE=growthmate-api/build/libs/growthmate-api-0.0.1-SNAPSHOT.jar
+COPY ${JAR_FILE} ./app.jar
+ENV TZ=Asia/Seoul
+ENTRYPOINT ["java","-jar","./app.jar"]
```
개행 추가해주세요!

ldk980130 commented on docker-compose-local.yml:
```diff
@@ -0,0 +1,16 @@
+      - ./database/config:/etc/mysql/conf.d
+      - ./database/init:/docker-entrypoint-initdb.d
+    ports:
+      - "3306:3306"
```
여기도 개행 추가 부탁해요!

ldk980130 commented on docker-compose-local.yml:
```diff
@@ -0,0 +1,16 @@
+version: "3.8"                                          # 파일 규격 버전
+services:                                               # 이 항목 밑에 실행하려는 컨테이너들을 정의
+  growthmate-database:
+    container_name: growthmate-database
+    build:
+      dockerfile: Dockerfile
+      context: ./database
+    image: mysql:latest
+    environment:
+      - MYSQL_DATABASE=growthmate
+      - MYSQL_ROOT_PASSWORD=1234
+    volumes:
+      - ./database/config:/etc/mysql/conf.d
+      - ./database/init:/docker-entrypoint-initdb.d
+    ports:
+      - "3306:3306"
```
로컬에서 mysql 실행하는 yml 같은데 로컬에서도 mysql로 실행하고 테스트해야할 필요가 있을까요? 매번 도커를 켜야 테스트 및 로컬 실행이 가능해질 것 같은데 H2 데이터베이스로 간단하게 로컬에서 돌리는 방법은 어떨까요?

Jeongh00 replied:
넵. 로그인 pull request 와 반복되는 리뷰인것같은데, h2 로 설정 완료하였습니다. docker compose 를 이용해 실제 상황에 맞게 잘 동작하는지 확인하고 싶어 mysql 로 실행했었는데 h2 를 이용하면 굳이 필요하지 않겠네요. 일단 파일은 남겨두겠습니다!

ldk980130 commented on docker-compose.yml:
```diff
@@ -0,0 +1,13 @@
+      - SPRING_PROFILES_ACTIVE=prod
+    ports:
+      - "80:8080"
+    restart: always
```
여기도 개행!

ldk980130 commented on growthmate-core/src/main/resources/db/migration/V1__init.sql:
```diff
@@ -17,7 +17,7 @@ CREATE TABLE post
     title      VARCHAR(255)          NOT NULL,
     content    LONGTEXT              NOT NULL,
-    is_deleted BIT(1)                NOT NULL,
+    is_deleted BIT                   NOT NULL,
```
이건 저도 무슨 차이인지는 잘 모르겠네요ㅠ

Jeongh00 replied:
Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Syntax error in SQL statement "CREATE TABLE comment\000a(\000a    comment_id BIGINT AUTO_INCREMENT NOT NULL,\000a    post_id    BIGINT                NOT NULL,\000a    member_id  BIGINT                NOT NULL,\000a    content    LONGTEXT              NOT NULL,\000a    is_deleted BIT[*](1)                NOT NULL,\000a    CONSTRAINT pk_comment PRIMARY KEY (comment_id)\000a)"; expected "ARRAY, INVISIBLE, VISIBLE, NOT NULL, NULL, AS, DEFAULT, GENERATED, ON UPDATE, NOT NULL, NULL, AUTO_INCREMENT, DEFAULT ON NULL, NULL_TO_DEFAULT, SEQUENCE, SELECTIVITY, COMMENT, CONSTRAINT, COMMENT, PRIMARY KEY, UNIQUE, NOT NULL, NULL, CHECK, REFERENCES, AUTO_INCREMENT, ,, )";

-> 그 때 이렇게 에러 떴었는데, 참고로 일단 보냅니다!

ldk980130 commented on growthmate-common/build.gradle:
```diff
@@ -6,3 +6,7 @@ dependencies {
     testFixturesImplementation 'org.springframework.boot:spring-boot-starter-data-jpa'
     testFixturesImplementation 'org.springframework.boot:spring-boot-starter-test'
 }
+
+springBoot {
+    mainClass = 'com.growup.growthmate.GrowthMateApiApplication'
+}
```
이것도 반드시 필요한 코드던가요? 이렇게 되면 common 모듈에서 api 모듈에 있는 클래스를 의존하게 되는데 이러면 양방향 의존이 되어버려요. api 모듈에서만 mainClass를 지정하고 common, core 모듈에 있는 gradle에는 mainClass 지정을 안하는 상태에선 스크립트가 실패하는지 궁금합니다.

Jeongh00 replied:
- > Failed to calculate the value of task ':growthmate-common:bootJar' property 'mainClass'.
-       > Main class name has not been configured and it could not be resolved from classpath build/classes/java/main

Jeongh00 replied:
저도 그렇게 하고 싶었는데 위와 같이 각각 모듈에 넣어주지 않으면 mainClass 를 못 찾겠다고 뜨네요...ㅠ 저번 플젝 경험때도 이런 경험이 있었는데 결국 어떤 문제인지 찾지 못했습니다. 무엇이 문제일까요..?

ldk980130 replied:
메인 클래스를 아무것도 지정하지 않은 상태에서 api 모듈만 빌드하면 빌드가 로컬에선 성공하긴 합니다. 이렇게 배포가 가능할까요?
`./gradlew growthmate-api:build`

저는 develop 브랜치에서 해본거긴합니다.

Jeongh00 replied:
제가 dockerfile을 api module jar 파일로 적어놔서 배포는 가능할것같은데 애초에 제가 develop 브랜치에서 빌드 했을떄 실패합니다. 

### Unsupported Java. 
Your build is currently configured to use Java 17.0.6 and Gradle 7.0.

Possible solution:
 - Use Java 16 as Gradle JVM: Open Gradle settings 
 - Upgrade Gradle wrapper to 7.2 version and re-import the project
 
 이렇게 에러가 뜨고있네요...ㅠ

ldk980130 commented on growthmate-core/settings.gradle:
```diff
@@ -0,0 +1,2 @@
+rootProject.name = 'growthmate'
+include 'growthmate-common'
```
개행 추가해주세요

ldk980130 reviewed (changes requested):
피드백보다는 질문이 많습니다. 답변 잘 부탁드립니다!

Jeongh00 committed:
feat: 개행 추가

Jeongh00 committed:
feat: cd submodule 추가 & 하위 모듈 mainClass 제거

Jeongh00 committed:
fix: 각 모듈별로 mainClass 다시 설정

Jeongh00 committed:
feat: gradle version up & mainclass 없이 빌드

Jeongh00 committed:
test

# <타입> : <제목> 의 형식으로 제목을 아래 공백줄에 작성
# 제목은 50자 이내 / 변경사항이 "무엇"인지 명확히 작성 / 끝에 마침표 금지
# 예) feat : 로그인 기능 추가

# 바로 아래 공백은 지우지 마세요 (제목과 본문의 분리를 위함)

################
# 본문(구체적인 내용)을 아랫줄에 작성
# 여러 줄의 메시지를 작성할 땐 "-"로 구분 (한 줄은 72자 이내)

################
# 꼬릿말(footer)을 아랫줄에 작성 (현재 커밋과 관련된 이슈 번호 추가 등)
# 예) Close #7

################
# feature : 새로운 기능 추가
# fix : 버그 수정
# docs : 문서 수정
# test : 테스트 코드 추가
# refact : 코드 리팩토링
# style : 코드 의미에 영향을 주지 않는 변경사항
# chore : 빌드 부분 혹은 패키지 매니저 수정사항
################

Jeongh00 committed:
feat: scp file dockercompose image 등록

Jeongh00 committed:
fix: file dir 변경

Jeongh00 committed:
fix: change username

Jeongh00 committed:
fix: file 이름 지정

Jeongh00 committed:
fix: strip-components 삭제

Jeongh00 committed:
fix: settings.gradle 파일 삭제

ldk980130 reviewed (approved):
수고하셨습니다! 2가지만 확인해주시고 merge하면 될 것 같습니다

1. dockerfile을 보니 prod 환경으로 실행하는 것 같진 않던데 배포된 버전이 prod 환경이 맞나요?
2. merge 하기 전에 브랜치 이름 main으로 바꾸고 merge해주세요!

Jeongh00 commented:
1. 넵 prod 환경 배포 상태인데, 혹시 dockerfile 에서 어떤 부분이 prod 환경이 아닌지 설명해주실 수 있나요? 이해를 잘 못했습니다.
2. 아 이미 merge 를 해버렸네요.. 혹시 feature 브랜치 이름을 main 으로 바꾸라는 뜻이었을까요? 그렇게 하는 이유를 혹시 설명해주실 수 있나요?
